### PR TITLE
Fix UB in SFTMap implementations

### PR DESCRIPTION
**Pull request by playXE**
1) Replaces casts of references to mutable references with `UnsafeCell<T>`. 
2) Spaces are stored inside each of implementation as pointers in order to make Rust compiler happy because of lifetimes.

### CODE REVIEW

**Comment by qinsoon** [comment]
> 1. Replaces casts of references to mutable references with `UnsafeCell<T>`.
> 2. Spaces are stored inside each of implementation as pointers in order to make Rust compiler happy because of lifetimes.

Thanks for the PR. The use of `UnsafeCell` looks fine as a workaround for the UB (eventually we would like to have a safe implementation). Anyway, it looks good. But I don't understand the second point -- is the new version of Rust compiler complaining about the lifetime? The current `SFTMap` implementation does not assume the lifetime of `&SFT`, and the creation of `SFTMap` states `&SFT` is `'static`. This allows us to easily change the scope of `SFTMap` in the future (e.g. we can confine the lifetime of `SFTMap` to the same lifetime as `MMTK`).

File: src/policy/sft_map.rs [hunk 1]
```diff
@@ -128,17 +134,19 @@ mod space_map {
                     assert!(start + bytes <= space_start + MAX_SPACE_EXTENT);
                 }
             }
-            *mut_self.sft.get_unchecked_mut(index) = space;
+
+            //*mut_self.sft.get_unchecked_mut(index) = space;
```
Comment (k-sareen) [inline src/policy/sft_map.rs:138]: Minor: Could you remove the commented out code here? I think there are a couple of other places too.

**Comment by playXE** [comment]
> > 1. Replaces casts of references to mutable references with `UnsafeCell<T>`.
> > 2. Spaces are stored inside each of implementation as pointers in order to make Rust compiler happy because of lifetimes.
> 
> Thanks for the PR. The use of `UnsafeCell` looks fine as a workaround for the UB (eventually we would like to have a safe implementation). Anyway, it looks good. But I don't understand the second point -- is the new version of Rust compiler complaining about the lifetime? The current `SFTMap` implementation does not assume the lifetime of `&SFT`, and the creation of `SFTMap` states `&SFT` is `'static`. This allows us to easily change the scope of `SFTMap` in the future (e.g. we can confine the lifetime of `SFTMap` to the same lifetime as `MMTK`).

The second point arises from the fact that `mut_self` is removed. Before it would create a reference to SFTMap implementation that had different lifetime and `space` that is passed to `ugprade` could get assigned to it, now when replaced with UnsaafeCell lifetimes do not match

File: src/policy/sft_map.rs (hunk 1 above)
Comment (playXE) [reply src/policy/sft_map.rs]: Should be fixed now

Comment (k-sareen) [reply src/policy/sft_map.rs]: Thanks. I think there are some other locations where the previous code has been commented out too. Could you remove those lines as well?

Comment (playXE) [reply src/policy/sft_map.rs]: Should be done now for sure

File: src/policy/sft_map.rs [hunk 2]
```diff
@@ -348,49 +348,51 @@ mod dense_chunk_map {
...
-        #[allow(clippy::cast_ref_to_mut)]
-        #[allow(clippy::mut_from_ref)]
-        unsafe fn mut_self(&self) -> &mut Self {
-            &mut *(self as *const _ as *mut _)
-        }
-
         pub fn addr_to_index(addr: Address) -> u8 {
             SFT_DENSE_CHUNK_MAP_INDEX.load_atomic::<u8>(addr, Ordering::Relaxed)
         }
+        #[inline(always)]
```
Comment (k-sareen) [inline src/policy/sft_map.rs:366]: Could you remove the `#[inline(always)]` here and for `index_map`? We've discovered that adding inline pragmas isn't always ideal and we prefer using profile-guided optimizations nowadays.

File: src/policy/sft_map.rs (hunk 1 above)
Comment (k-sareen) [reply src/policy/sft_map.rs]: Thanks!

File: src/policy/sft_map.rs (hunk 2 above)
Comment (playXE) [reply src/policy/sft_map.rs]: Done

Comment (k-sareen) [reply src/policy/sft_map.rs]: Thanks :)

**Comment by playXE** [comment]
Hmm I would not merge this PR for now, I'm going to make storage of pointers in SFT vectors Atomic as well, using Atomic<T> from [atomic](https://docs.rs/atomic)

**Comment by playXE** [comment]
> Hmm I would not merge this PR for now, I'm going to make storage of pointers in SFT vectors Atomic as well, using Atomic from [atomic](https://docs.rs/atomic)

So far @qinsoon told me that update method is invoked behind the lock so probably `push()` and update of spaces inside it are fine without atomic pointers. In that case merging this PR should be fine.

**Comment by qinsoon** [comment]
> > Hmm I would not merge this PR for now, I'm going to make storage of pointers in SFT vectors Atomic as well, using Atomic from [atomic](https://docs.rs/atomic)
> 
> So far @qinsoon told me that update method is invoked behind the lock so probably `push()` and update of spaces inside it are fine without atomic pointers. In that case merging this PR should be fine.

Can you fix the style check? Then I will start a test run with our bindings to test the PR, then we can merge it.

**Comment by playXE** [comment]
> > > Hmm I would not merge this PR for now, I'm going to make storage of pointers in SFT vectors Atomic as well, using Atomic from [atomic](https://docs.rs/atomic)
> > 
> > 
> > So far @qinsoon told me that update method is invoked behind the lock so probably `push()` and update of spaces inside it are fine without atomic pointers. In that case merging this PR should be fine.
> 
> Can you fix the style check? Then I will start a test run with our bindings to test the PR, then we can merge it.

Hmm, weirdly enough I don't get the same error when I git clone it and run cargo fmt check and clippy, could you maybe restart pre code review check?

**Review by qinsoon — APPROVED** [review]
LGTM

**Pull request merged**